test/spanningtrees/wilson.jl

```diff
@@ -0,0 +1,84 @@
+#test that the result of random_spanning_tree is correct
+#ie. it should be a spanning forest, oriented towards the roots 
+function check_correctness(edges,roots)
+    F = SimpleDiGraph(edges)
+    nmissing = max(0,Int.(maximum(roots))-nv(F)) #there might be some missing nodes if they're alone in their components
+    add_vertices!(F,nmissing)
+    @test !is_cyclic(F)
+    @test all(outdegree(F,roots) .== 0)
+    #test that all nodes lead to a root
+    for cc in connected_components(F)
+        if (length(cc)==1)
+            @test cc[1] in roots
+        else
+            rc = intersect(cc,roots)
+            #there should be a single root per connected component
+            @test length(rc)==1
+            rc = rc[1]
+            bf=bfs_parents(F,rc;dir=-1)[cc]
+            @test all(bf .> 0)
+        end
+    end
+end
+#rather than import from StatsBase
+function relcounts(values :: Array{Int,1},k :: Int)
+    cc=zeros(k)
+    for v in values
+        cc[v] += 1
+    end
+    cc ./ length(values)
+end
+
+@testset "Wilson" begin
+    gg = Grid([4,4])
+    for g in testgraphs(gg)
+        rt = @inferred(random_spanning_tree(g))
+        check_correctness(rt.edges,rt.roots)
+    end
+
+    #Try a bunch of random graphs
+    rg = [SimpleGraph(20,20) for i in 1:15]
+    map((g) -> (rt=random_spanning_tree(g);check_correctness(rt.edges,rt.roots)),rg)
+
+    #The next graph is not connected, a forest with three roots should be returned
+    G=reduce(blockdiag,[CycleDiGraph(5),CycleDiGraph(3),CycleDiGraph(2)])
+    rt=random_spanning_tree(G)
+    check_correctness(rt.edges,rt.roots)
+    @test length(rt.roots)==3
+
+    # What follows is a probabilistic test that checks that the algorithm samples from the correct distribution
+    # We use a result by Pemantle & Wilson that states that
+    # Uniform Spanning Trees are actually a Determinantal Point Process over the edges of the graph
+    # The kernel of the DPP can be computed explictly from the edge-incidence matrix
+    IM = Matrix(incidence_matrix(gg;oriented=true));
+    eg = eigen(IM'*IM);
+    valid = abs.(eg.values) .> 1e-12
+    U = eg.vectors[:,valid];
+    K = U*U';
+    # The kernel of the DPP gives the inclusion probabilities: for instance K_ii is the probability that edge i is included in a UST
+    # We compare theoretical to observed incl. probabilities 
+    AA = [adjacency_matrix(SimpleGraph(random_spanning_tree(gg).edges)) for i in 1:100000];
+    prob_incl = reduce(+,AA)./length(AA);
+    pr = map((e) -> prob_incl[src(e),dst(e)],edges(gg));
+    #Check that observed probabilities don't deviate too much 
```

Comment:
Avoid nondeterministic tests here. What is it that you want to ensure you cover, and can you do that with a few selected graphs?

Reply:
Here the spirit of the test isn't to try that a certain property holds across a number of random graphs, it's rather that a certain property should hold for a given graph, over realisations of the algorithm. Specifically, for spanning trees sampled uniformly, we can explicitly compute the probability that edge e belongs to the sampled tree. The test checks that the algorithm's output verify the theoretical sampling probabilities, give or take random error. 
It's hard to see how to do this in a non-Monte Carlo manner